README.md
# tentopod

[![Version](http://cocoapod-badges.herokuapp.com/v/tentopod/badge.png)](http://cocoadocs.org/docsets/tentopod)
[![Platform](http://cocoapod-badges.herokuapp.com/p/tentopod/badge.png)](http://cocoadocs.org/docsets/tentopod)

## Requirements

Before you can start using the push service you'll have to register your App at APNS first. You'll get a guide on how to do this with your 10to1 app data. 

## Installation

tentopod is available through [CocoaPods](http://cocoapods.org), to install
it simply add the following line to your Podfile:

    pod "tentopod"
To install run:	

	pod install

## Usage

Before you can use the 10topod library in your Xcode project, you first have to import the public header. Place the following line between your imports in your AppDelegate.m file:

	#import <TTOTenToPush.h>
Now the real fun starts. We'll start by registering your device to receive push notifications. Add the following block of code to your didFinishLaunchingWithOptions method. BE CAREFUL! Replace the IDENTIFIER and TOKEN constants with the data 10to1 gave you or the service won't work.

	[[TTOTenToPush shared]setupWithIdentifier:IDENTIFIER token:TOKEN];
    [[TTOTenToPush shared]registerTenToPush:[UIApplication sharedApplication]];
This will set up the push service to receive push messages the default way. For the more adventurous developers: We also built in a method to customize your push messages. Just pass on your default iOS notification badges as demonstrated below:

	[[TenToPush shared ] registerTenToPush: [UIApplication sharedApplication] withInitializers:(UIRemoteNotificationTypeBadge | UIRemoteNotificationTypeSound | UIRemoteNotificationTypeAlert)];
This will do pretty much all the magic so you're almost done. There's just one more thing you have to do if you want to be able to send push messages to the client device. Add the top method to your AppDelegate.m file. This one will make sure that the devices get registered on the 10to1 webserver where the push messages are sent from. The second method is useful for error handling. 

	- (void)application:(UIApplication*)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData*)deviceToken
	{
	    [[TTOTenToPush shared] registerDeviceToken: deviceToken];
    
	}
	
	- (void)application:(UIApplication*)application didFailToRegisterForRemoteNotificationsWithError:(NSError*)error
	{
		NSLog(@"Failed to get token, error: %@", error);
	}
These two methods are not essential for the service to work but they're sure useful if you want to do a bit more with your app than just receive push messages. getUniqueApid returns your unique 10to1 device identifier, getDeviceToken returns your APNS registration ID.

    [[TTOTenToPush shared] getUniqueApid ];
    [[TTOTenToPush shared] getDeviceToken ];


## Author

Gregory Lambrechts, lambrechts_gregory@hotmail.com

## License

tentopod is available under the MIT license. See the LICENSE file for more info.

=======
TenToPod
========

Development version of the iOS library for the 10topush service.


CHANGELOG.md
# tentopod CHANGELOG

## 0.1.0

Initial release.


tentolib/tentolib/TTOInstall.m
//
//  Install.m
//  iostentopushsdk
//
//  Created by 10to1 on 16/04/14.
//  Copyright (c) 2014 10to1. All rights reserved.
//

#import "TTOInstall.h"
#import <SVHTTPRequest.h>

@implementation TTOInstall

-(id) init
{
    self = [super init];
    if (self)
    {
        // superclass successfully initialized, further
        // initialization happens here ...
    }
    return self;
}

//Post to Installs API.
- (NSDictionary*)jsonDictionary:(NSString *) deviceToken packageName:(NSString*)packageName secretKey:(NSString *)secretKey{
    NSDictionary* out;
    NSDictionary* params;
    
    out = @{ @"package_name" : packageName, @"secret_key" : secretKey, @"install" : params =@{
                                                            @"type" : @"IosInstall",
                                                            @"unique_apid" : @"",
                                                            @"device_token" : deviceToken,}};
    
    return out;
}



@end



tentolib/tentolib/TTOTenToPush.m
//
//  TenToPush.m
//  iostentopushsdk
//
//  Created by 10to1 on 16/04/14.
//  Copyright (c) 2014 10to1. All rights reserved.
//

#import "TTOTenToPush.h"
#import <UIKit/UIApplication.h>
#import "TTOInstall.h"
#import <SVHTTPRequest.h>

@implementation TTOTenToPush

+ (instancetype)shared
{
    static dispatch_once_t once;
    static id sharedInstance;
    dispatch_once(&once, ^{
        sharedInstance = [self new];
    });
    return sharedInstance;
}


//Default app configurations. Package name + secret key received from 10to1.
- (void)setupWithIdentifier:(NSString *)packageName token:(NSString *)secret{
    [self setPackageName:packageName];
    [self setSecretKey:secret];
    [self setInstall:[[TTOInstall alloc] init]];
        
}

// Register with tokens.
- (void) registerTenToPush:(UIApplication *)application withInitializers:(UIRemoteNotificationType) types{
    [application registerForRemoteNotificationTypes: types];
}

//Default way of registering.
- (void) registerTenToPush:(UIApplication *)application{
    [application registerForRemoteNotificationTypes:(UIRemoteNotificationTypeBadge | UIRemoteNotificationTypeSound | UIRemoteNotificationTypeAlert)];
}

//Register device token with 10to1 push server.
- (void)registerDeviceToken:(NSString *) deviceToken{
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSString *uniqueApid = [defaults objectForKey:@"deviceToken"];
    if(uniqueApid == nil){
        NSString *preparedDeviceToken = [self prepareDeviceToken:deviceToken];
        [defaults setObject:preparedDeviceToken forKey:@"deviceToken"];
        [defaults synchronize];
        NSLog(@"Data saved");
        NSLog(@"devicetoken");
        [self sendToBackEnd:preparedDeviceToken];
    }
}

//Delete all whitespaces and brackets from the device token for clean send to server.
- (NSString *) prepareDeviceToken: (NSString *)deviceToken
{
    NSString *dt = [deviceToken description];
    NSString *whiteSpaceString = [dt stringByReplacingOccurrencesOfString:@" " withString:@""];
    NSString *first = [whiteSpaceString substringFromIndex:1];
    NSString *out = [first substringToIndex:first.length-(first.length>0)];
    return out;
}

//Returns device token.
- (NSString *) deviceToken{
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    
    NSString *otherdevicetoken = [defaults objectForKey:@"deviceToken"];
    return otherdevicetoken;
}

//Returns unique apid.
- (NSString *) uniqueApid{
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSString *apid = [defaults objectForKey:@"uniqueApid"];
    return apid;
    
}

//Post to Installs API.
- (void) sendToBackEnd:(NSString *) deviceToken{
    
    NSLog(@"backend");
    
    [[SVHTTPClient sharedClient] setBasePath:@"http://push.maggie.10to1apps.be/"];
    
    [SVHTTPClient sharedClient].sendParametersAsJSON = YES;
    
    [[SVHTTPClient sharedClient] POST:@"api/installs" parameters: [[self install] jsonDictionary:deviceToken packageName:[self packageName] secretKey:[self secretKey]]
                           completion:^(id response, NSHTTPURLResponse *urlResponse, NSError *error) {
                               if(response != nil)
                               {
                                   NSLog(@"%@",response);
                                   NSString *apid =  [response valueForKey:@"unique_apid"];
                                   NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
                                   [defaults setObject:apid forKey:@"uniqueApid"];
                                   [defaults synchronize];
                               }
                               else
                               {
                                    #ifdef DEBUG
                                    #define MyLog(str) NSLog(str,error);
                                    #endif
                               }
                           }];
    
}


@end
